Elastic indexer

AvinashBukkittu opened the pull request:
This PR

- Adds evaluator support for the pipeline. This means, we can now add an evaluator to the pipeline and call `evaluate()` on the pipeline to evaluate on a dataset.
- Adds ElasticIndexer along with `ElasticSearchIndexProcessor` processor to index the documents.
- Adds ElasticSearchProcessor for searching documents in an elastic indexer
- Creates Passage Reranker example for MS Marco Dataset. Provides a baseline model for ranking using just Elastic Indexer
  - Adds an EvalReader to read MS Marco eval dataset
  - Adds MS Marco Eval script in the example

hunterhector commented on forte/processors/elastic_search_query_creator.py:
```diff
@@ -0,0 +1,57 @@
+            "query_pack_name": "query"
+        }
+
+    def _process_query(self, input_pack: MultiPack):
```
Return type would help

AvinashBukkittu replied:
Added in 4776d40

hunterhector commented on forte/processors/elastic_search_processor.py:
```diff
@@ -0,0 +1,72 @@
+
+    def _process(self, input_pack: MultiPack):
+        query_pack = input_pack.get_pack(self.config.query_pack_name)
+        first_query = list(query_pack.get_entries(Query))[0]
```
It would help future developers if we comment why we use the first query here.

AvinashBukkittu replied:
Added in 4776d40

hunterhector commented on forte/processors/elastic_search_processor.py:
```diff
@@ -0,0 +1,72 @@
+            # TODO: add the BM score and the labels for retrieved documents
+            # which will be used by the reranker
+            # Convert `Document` to `Passage`
+            document = Document(pack=pack, begin=0, end=len(content))
```
This document entry doesn't not have meta data apart from the character range, what is its utility?

AvinashBukkittu replied:
Each retrieved result is added as `Document` type in the new pack. It is similar to how `SearchProcessor` was adding `Document` types. If the subsequent processors need to fetch the results, they can use `get_entries_by_type(...)` on this pack.

hunterhector commented on forte/processors/base/index_processor.py:
```diff
@@ -0,0 +1,63 @@
+]
+
+
+class IndexProcessor(BaseProcessor[DataPack], ABC):
```
Let's have a class docstring here.

AvinashBukkittu replied:
Added in 4776d40

hunterhector commented on examples/passage_reranker/ms_marco_evaluator.py:
```diff
@@ -0,0 +1,39 @@
+from texar.torch import HParams
```
Well, header is missing here.

AvinashBukkittu replied:
Added in 2fc3e5f

hunterhector commented on examples/passage_reranker/reader.py:
```diff
@@ -0,0 +1,38 @@
+from typing import Iterator
```
Header

AvinashBukkittu replied:
Added in 2fc3e5f

hunterhector commented on forte/indexers/indexers.py:
```diff
@@ -37,20 +40,147 @@ class BaseIndexer(ABC):
+        index_name = index_name if index_name else self.hparams.index_name
+        return self.elasticsearch.search(index=index_name, body=query)
+
+    def search_all(self, query: Dict[str, Any]):
```
What's the plan for this method? If we are not planning to implement this soon, maybe we should add not_implemented error here?

AvinashBukkittu replied:
Removed it in b7632b8

hunterhector commented on forte/processors/elastic_search_processor.py:
```diff
@@ -0,0 +1,72 @@
+    def _process(self, input_pack: MultiPack):
+        query_pack = input_pack.get_pack(self.config.query_pack_name)
+        first_query = list(query_pack.get_entries(Query))[0]
+        results = self.index.search(first_query.value)
```
Now the design for `index.search` is that it takes the value of `first_query`. How about taking the whole `query` entry? That way if we have a more complex query we can represent them with the `query` object, instead of forcing everything inside `value`.

AvinashBukkittu replied:
Hmm, the `value` attribute of `Query` was designed to be generic enough to hold any complex query. `Query` object may hold other attributes like `rank_list` etc which maybe irrelevant to the indexer.

AvinashBukkittu commented:
Importing relevant comments from https://github.com/asyml/forte/pull/103

1) https://github.com/asyml/forte/pull/103#discussion_r361890259
> Let's have more typing here.

Added typing in MS Marco Evaluator

2) https://github.com/asyml/forte/pull/103#discussion_r361890348
> It would be better to store only some necessary information from the pack, here we only need the doc_id?

Simplified the logic of MS Marco Evaluator in b7632b8

3) https://github.com/asyml/forte/pull/103#discussion_r361890717
> passage is too specific as a name. how about rank_list?

Changed to `results` from `passages`

4) https://github.com/asyml/forte/pull/103#discussion_r361891432
> Add some docstring here to teach users to extend this method in order to create more complex queries.

Done in b7632b8

5) https://github.com/asyml/forte/pull/103#discussion_r361891867
> Do we need to benchmark the speed of the indexer? Hopefully, our wrapper won't decrease the speed a lot.

Added a benchmark testcase in b7632b8

6) https://github.com/asyml/forte/pull/103#discussion_r361892701
> The design of the _process_query seem to be a little difficult, especially on returning input_pack.

I agree. Essentially, the following three lines

```
query = Query(pack=query_pack)
query.set_value(value=query_value)
query_pack.add_entry(query)
``` 
are common in `QueryProcessor`. If we want to abstract away these details in `_process` method, I couldn't think of a better way other than to return the `query_pack` and the `query_value` from `_process_query`

7) https://github.com/asyml/forte/pull/103#discussion_r361892941
> at least in bulk mode, we should add a couple more documents.

Increased the limit to 10,000
